oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoder.java

```diff
@@ -309,6 +336,38 @@ private static URL toURL(String url) {
 			}
 		}
 
+		private static class NoOpJwkSetCache implements JWKSetCache {
+			@Override
+			public void put(JWKSet jwkSet) {
+			}
+
+			@Override
+			public JWKSet get() {
+				return null;
+			}
+
+			@Override
+			public boolean requiresRefresh() {
+				return true;
+			}
+		}
+
+		private static class CachingResourceRetriever implements ResourceRetriever {
+			private final Cache cache;
+			private final ResourceRetriever resourceRetriever;
+
+			CachingResourceRetriever(Cache cache, ResourceRetriever resourceRetriever) {
+				this.cache = cache;
+				this.resourceRetriever = resourceRetriever;
+			}
+
+			@Override
+			public Resource retrieveResource(URL url) {
+				String jwkSet = cache.get(url.toString(), () -> resourceRetriever.retrieveResource(url).getContent());
```

Comment:
I think we may want to wrap any exception in an `IOException`, like `RestOperationsResourceRetriever` does. The reason for that is so that Nimbus's `RemoteJWKSet` can react appropriately and throw its `RemoteKeySourceException`. 

My concern is that adding this cache will change the exception callers see from a `RemoteKeySourceException` to a `ValueRetreivalException`.